### DISCUSSION TITLE
AutoTDP algorithm and options

### DISCUSSION THREAD

**Discussion by buri-92** [seerge/g-helper#2993] [Ideas]
Hey man, first of all I am very grateful for you work and I hopefully don't come across as rude or greedy, but I wanted to ask/suggest if it would be possible to expand on the AutoTDP feature and its algorithm? Maybe offer profiles (efficiency vs gameplay) or give users options for customization?

I was looking at the source code (unfortunately I am not that familiar with C#) but if I understood correctly every 300 Milliseconds the Timer_Elapsed method with the main logic of AutoTDP is called, so the check for TDP adjustments occurs every 300 ms?

And here the TDP is increased by 1 unit (Watt?) if the FPS is < 90% of the target FPS, and decreased by 1 unit if current FPS is > 95% target FPS for 8 cycles, correct?

As someone who likes to play with 40 or 45 fps on my Ally to maximize battery life, drops from 40 to < 36 FPS or 45 to < 41 FPS feel very noticeable and it takes a little time until the TDP has adjusted, which leads to very uneven frame times and unfun gameplay in usually the most hectic situations.
So from a gameplay perspective the minimum TDP check is, at least for me, nearly not aggressive enough. Especially in games that run great with a lot of headroom and suddenly throw those curveballs at you (looking at you Cyberpunk...)

If I understood your code correctly, the 3 main elements of your algorithm are the check interval, the minimum FPS percentage and the TDP increase step.
_I would love it if you could offer options for users to customize those 3 settings ourselves._
Especially for the lower FPS limits, it would be crucial to a better gaming experience if those 3 values were a lot more aggressive.

---

From this point I am just throwing ideas out there:

For example, unless it is too much stress on the system, I wondered why not match the check interval with the frame time of the target fps? So, 30 fps => 33.3 ms, 40 fps => 25 ms, 45 fps => 22.2 ms etc.. 
Maybe this would lead to too many calls for higher fps targets and is probably also unnecessary for those, dropping from 90 fps to 80 is a lot less bothersome after all, but for the lower fps limits I think a reaction time of only 1 frame is very welcome.

Also, on top of a more aggressive minFPS % (I would go at least 95%, at least for 30, 40 and 45 fps), I think the TDP increase step should be dynamic, maybe taking into account the GPU load and definitely be relative to the current FPS/target FPS. So if for some reason the game suddenly drops from 40 FPS to 30 FPS and GPU load jumps from 80% to >95%, the TDP should in my opinion instantly increase to the max. limit set by the user to ensure smoother gameplay. 
Maybe something like newTDP = currentTDP * ( targetFPS / currentFPS ) (not a math guy...) would work, increasing the TDP by the % the current FPS is away from the target? The closer we are to the target, the smaller the increase.

If the customizable options are too much work, I would love it if I was able to convince you at least of some of my ideas.

Again, thanks very much for your work and sorry for the long text, lol.

**Comment by seerge** [comment]
@buri-92 hello, whole AutoTDP code is here
https://github.com/seerge/g-helper/blob/cd930cebf9585c5999e417884c78fdfa305afbf0/app/Ally/AllyControl.cs#L351

It's already tweaked to be very soft (i.e. not aggressive) and "safe" and aims at 90..95% from target FPS, doing measurements and adjustments every 300ms.

- If current FPS <= 90% **it immediately adds 1W of power**,  
- If FPS >= 95% over next 8 measurements (2.4 seconds) **AND there is at least 1W of unused power** (for example TDP is 10W and APU uses 9W) it will deduct 1W of power

As you see criteria for going down in power is much more strict, so app will lower limits only when it's indeed a very stable FPS and there is a headroom for doing that

So if you have 40FPS limit, it aims at 36..38 FPS. I really doubt it's doable to keep range even more narrow than 2 FPS.

--

When app requests FPS counter from AMD driver, it reports **average FPS since last query**. If it will check every 30ms, it will be just lots of inaccurate measurements (due to unstable FPS nature in general).

Also due to the way power limits work in general, they can take like 500ms or even 1 second to actually "apply" on firmware level sometimes. So making frequent adjustments just makes no sense anyway.
